docs/features/interceptors.md

```diff
@@ -0,0 +1,121 @@
+# Interceptors
+
+## Summary
+[summary]: #summary
+
+*Interceptors* are an experimental compiler feature. An *interceptor* is a method which can declaratively substitute a call to itself instead of a call to an *interceptable* method at compile time. This substitution occurs by having the interceptor declare the source locations of the calls that it intercepts. This provides a limited facility to change the semantics of existing code by adding new code to a compilation (e.g. in a source generator).
+
+```cs
+using System;
+using System.Runtime.CompilerServices;
+
+var c = new C();
+c.InterceptableMethod(1); // (L1,C1): prints `interceptor 1`
+c.InterceptableMethod(1); // (L2,C2): prints `other interceptor 1`
+c.InterceptableMethod(1); // prints `interceptable 1`
+
+class C
+{
+    [Interceptable]
+    public void InterceptableMethod(int param)
+    {
+        Console.WriteLine($"interceptable {param}");
+    }
+}
+
+// generated code
+static class D
+{
+    [InterceptsLocation("Program.cs", line: /*L1*/, character: /*C1*/)] // refers to the call at (L1, C1)
+    public static void InterceptorMethod(this C c, int param)
+    {
+        Console.WriteLine($"interceptor {param}");
+    }
+
+    [InterceptsLocation("Program.cs", line: /*L2*/, character: /*C2*/)] // refers to the call at (L2, C2)
+    public static void OtherInterceptorMethod(this C c, int param)
+    {
+        Console.WriteLine($"other interceptor {param}");
+    }
+}
+```
+
+## Detailed design
+[design]: #detailed-design
+
+### InterceptableAttribute
+
+A method must indicate that its calls can be *intercepted* by including `[Interceptable]` on its declaration.
+
+If a call is intercepted to a method which lacks this attribute, a warning is reported. This may be changed to an error in the future.
```

Comment:
The call is still intercepted. I'll clarify that in the text.

This allows us to experiment with the scenarios we're interested in without having to annotate and rebuild the related libraries first.

We might make this an error before feature merge. It might depend on what our partners need from us in the short term.